source/extensions/transport_sockets/tls/cert_validator/spiffe/spiffe_validator.cc

```diff
@@ -0,0 +1,260 @@
+
+SPIFFEValidator::SPIFFEValidator(const Envoy::Ssl::CertificateValidationContextConfig* config,
+                                 SslStats& stats, TimeSource& time_source)
+    : stats_(stats), time_source_(time_source) {
+  ASSERT(config != nullptr);
+
+  SPIFFEConfig message;
+  Config::Utility::translateOpaqueConfig(config->customValidatorConfig().value().typed_config(),
+                                         ProtobufWkt::Struct(),
+                                         ProtobufMessage::getStrictValidationVisitor(), message);
+
+  auto size = message.trust_domains().size();
+  trust_bundle_stores_.reserve(size);
+  for (auto& domain : message.trust_domains()) {
+    if (trust_bundle_stores_.find(domain.name()) != trust_bundle_stores_.end()) {
+      throw EnvoyException(absl::StrCat(
+          "Multiple trust bundles are given for one trust domain for ", domain.name()));
+    }
+
+    auto cert = Config::DataSource::read(domain.trust_bundle(), true, config->api());
+    bssl::UniquePtr<BIO> bio(BIO_new_mem_buf(const_cast<char*>(cert.data()), cert.size()));
+    RELEASE_ASSERT(bio != nullptr, "");
+    bssl::UniquePtr<STACK_OF(X509_INFO)> list(
+        PEM_X509_INFO_read_bio(bio.get(), nullptr, nullptr, nullptr));
+    if (list == nullptr || sk_X509_INFO_num(list.get()) == 0) {
+      throw EnvoyException(
+          absl::StrCat("Failed to load trusted CA certificate for ", domain.name()));
+    }
+
+    auto store = X509StorePtr(X509_STORE_new());
+    bool has_crl = false;
+    bool ca_loaded = false;
+    for (const X509_INFO* item : list.get()) {
+      if (item->x509) {
+        X509_STORE_add_cert(store.get(), item->x509);
+        X509_up_ref(item->x509);
+        ca_certs_.push_back(bssl::UniquePtr<X509>(item->x509));
+        if (!ca_loaded) {
+          // TODO: With the current interface, we cannot return the multiple
+          // cert information on getCaCertInformation method.
+          // So temporarily we return the first CA's info here.
+          ca_loaded = true;
+          ca_file_name_ = absl::StrCat(domain.name(), ": ",
+                                       domain.trust_bundle().filename().empty()
+                                           ? "<inline>"
+                                           : domain.trust_bundle().filename());
+        }
+      }
+
+      if (item->crl) {
+        has_crl = true;
+        X509_STORE_add_crl(store.get(), item->crl);
+      }
+    }
+    if (has_crl) {
+      X509_STORE_set_flags(store.get(), X509_V_FLAG_CRL_CHECK | X509_V_FLAG_CRL_CHECK_ALL);
+    }
+    trust_bundle_stores_[domain.name()] = std::move(store);
+  }
+}
+
+void SPIFFEValidator::addClientValidationContext(SSL_CTX* ctx, bool) {
+  bssl::UniquePtr<STACK_OF(X509_NAME)> list(sk_X509_NAME_new(
+      [](const X509_NAME** a, const X509_NAME** b) -> int { return X509_NAME_cmp(*a, *b); }));
+
+  for (auto& ca : ca_certs_) {
+    X509_NAME* name = X509_get_subject_name(ca.get());
```

Comment:
I'd like to understand why we enforce a unique set based on the subject name of the CA certificates? There is no constraint in the SPIFFE specification that X.509 authorities in a trust bundle have unique subject names. In fact, in practice, SPIRE server CAs part of the same deployment often have the same subject name for the CA certificates. Is there a particular situation we're trying to avoid here?

Reply:
This logic is what I borrowed from the existing certificate validation logic in Envoy, and not sure why it does deduplication on SAN. It's been 3+ years since [the PR was merged](https://github.com/envoyproxy/envoy/pull/2248). I guess this is only for `SSL_CTX_set_client_CA_list` and the openssl doc `The CAs listed do not become trusted (list only contains the names, not the complete certificates)`. 

cc @PiotrSikora since you are the author of the PR and an expert in this field, and could you give me the reason why the list for `SSL_CTX_set_client_CA_list` is deduplicated on SAN names?

Reply:
`SSL_CTX_set_client_CA_list` sets only a list of names, not a list of trusted CAs.

This function (well, not this one exactly, but the one this is a copy of) is a non-file based re-implementation of [`SSL_load_client_CA_file`](https://github.com/google/boringssl/blob/0a86cbff452f3c1807738968a88332416279e3dc/ssl/ssl_file.cc#L167)  which contains the same logic.

Reply:
Thanks, that makes sense. I believe this is fine as it is, so I'm marking this comment as resolved.

Reply:
I see now. Thanks for the answer!